File: day11/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
)

const (
	puzzleInput = "input.txt"
)

type (
	Stack struct {
		k []Pos
	}
)

func (s *Stack) Push(c Pos) {
	s.k = append(s.k, c)
}

func (s *Stack) Pop() (Pos, bool) {
	l := len(s.k)
	if l == 0 {
		return Pos{}, false
	}
	v := s.k[l-1]
	s.k = s.k[:l-1]
	return v, true
}

func (s *Stack) Peek() (Pos, bool) {
	l := len(s.k)
	if l == 0 {
		return Pos{}, false
	}
	return s.k[l-1], true
}

type (
	Pos struct {
		x, y int
	}

	Grid struct {
		rows [][]int
		w    int
		h    int
	}
)

func NewGrid(rows [][]int) *Grid {
	w := 0
	h := len(rows)
	if h > 0 {
		w = len(rows[0])
	}
	return &Grid{
		rows: rows,
		w:    w,
		h:    h,
	}
}

func (g *Grid) Step() int {
	flashes := map[Pos]struct{}{}
	stack := Stack{}
	for r, i := range g.rows {
		for c := range i {
			g.rows[r][c]++
			if g.rows[r][c] > 9 {
				k := Pos{
					x: c,
					y: r,
				}
				flashes[k] = struct{}{}
				stack.Push(k)
			}
		}
	}
	for {
		i, ok := stack.Pop()
		if !ok {
			break
		}
		if g.incr(i.x-1, i.y, flashes) {
			k := Pos{
				x: i.x - 1,
				y: i.y,
			}
			flashes[k] = struct{}{}
			stack.Push(k)
		}
		if g.incr(i.x-1, i.y-1, flashes) {
			k := Pos{
				x: i.x - 1,
				y: i.y - 1,
			}
			flashes[k] = struct{}{}
			stack.Push(k)
		}
		if g.incr(i.x, i.y-1, flashes) {
			k := Pos{
				x: i.x,
				y: i.y - 1,
			}
			flashes[k] = struct{}{}
			stack.Push(k)
		}
		if g.incr(i.x+1, i.y-1, flashes) {
			k := Pos{
				x: i.x + 1,
				y: i.y - 1,
			}
			flashes[k] = struct{}{}
			stack.Push(k)
		}
		if g.incr(i.x+1, i.y, flashes) {
			k := Pos{
				x: i.x + 1,
				y: i.y,
			}
			flashes[k] = struct{}{}
			stack.Push(k)
		}
		if g.incr(i.x+1, i.y+1, flashes) {
			k := Pos{
				x: i.x + 1,
				y: i.y + 1,
			}
			flashes[k] = struct{}{}
			stack.Push(k)
		}
		if g.incr(i.x, i.y+1, flashes) {
			k := Pos{
				x: i.x,
				y: i.y + 1,
			}
			flashes[k] = struct{}{}
			stack.Push(k)
		}
		if g.incr(i.x-1, i.y+1, flashes) {
			k := Pos{
				x: i.x - 1,
				y: i.y + 1,
			}
			flashes[k] = struct{}{}
			stack.Push(k)
		}
	}
	for i := range flashes {
		g.rows[i.y][i.x] = 0
	}
	return len(flashes)
}

func (g *Grid) incr(x, y int, flashes map[Pos]struct{}) bool {
	if g.outBounds(x, y) {
		return false
	}
	g.rows[y][x]++
	if g.rows[y][x] > 9 {
		_, ok := flashes[Pos{x: x, y: y}]
		return !ok
	}
	return false
}

func (g *Grid) outBounds(x, y int) bool {
	return x < 0 || y < 0 || x >= g.w || y >= g.h
}

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatal(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatal(err)
		}
	}()

	var rows [][]int
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := []byte(scanner.Text())
		row := make([]int, 0, len(line))
		for _, i := range line {
			row = append(row, int(i)-'0')
		}
		rows = append(rows, row)
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	grid := NewGrid(rows)
	count := 0
	for i := 0; i < 100; i++ {
		count += grid.Step()
	}
	fmt.Println("Part 1:", count)
	total := grid.w * grid.h
	step := 100
	for {
		step++
		if grid.Step() == total {
			fmt.Println("Part 2:", step)
			return
		}
	}
}

File: day11/src/main.rs
use std::collections::HashSet;
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";

#[derive(PartialEq, Eq, Hash, Clone, Copy)]
struct Pos(usize, usize);

struct Grid {
    rows: Vec<Vec<i32>>,
    w: usize,
    h: usize,
}

impl Grid {
    fn new(rows: Vec<Vec<i32>>) -> Self {
        let h = rows.len();
        let w = if h > 0 { rows[0].len() } else { 0 };
        Self { rows, w, h }
    }

    fn step(&mut self) -> usize {
        let mut flashes = HashSet::new();
        let mut queue = Vec::new();
        for i in 0..self.h {
            for j in 0..self.w {
                self.rows[i][j] += 1;
                if self.rows[i][j] > 9 {
                    let k = Pos(j, i);
                    flashes.insert(k);
                    queue.push(k);
                }
            }
        }
        loop {
            let Pos(x, y) = match queue.pop() {
                Some(i) => i,
                None => break,
            };
            if self.incr(x.wrapping_sub(1), y, &flashes) {
                let k = Pos(x.wrapping_sub(1), y);
                flashes.insert(k);
                queue.push(k);
            }
            if self.incr(x.wrapping_sub(1), y.wrapping_sub(1), &flashes) {
                let k = Pos(x.wrapping_sub(1), y.wrapping_sub(1));
                flashes.insert(k);
                queue.push(k);
            }
            if self.incr(x, y.wrapping_sub(1), &flashes) {
                let k = Pos(x, y.wrapping_sub(1));
                flashes.insert(k);
                queue.push(k);
            }
            if self.incr(x + 1, y.wrapping_sub(1), &flashes) {
                let k = Pos(x + 1, y.wrapping_sub(1));
                flashes.insert(k);
                queue.push(k);
            }
            if self.incr(x + 1, y, &flashes) {
                let k = Pos(x + 1, y);
                flashes.insert(k);
                queue.push(k);
            }
            if self.incr(x + 1, y + 1, &flashes) {
                let k = Pos(x + 1, y + 1);
                flashes.insert(k);
                queue.push(k);
            }
            if self.incr(x, y + 1, &flashes) {
                let k = Pos(x, y + 1);
                flashes.insert(k);
                queue.push(k);
            }
            if self.incr(x.wrapping_sub(1), y + 1, &flashes) {
                let k = Pos(x.wrapping_sub(1), y + 1);
                flashes.insert(k);
                queue.push(k);
            }
        }
        for &Pos(x, y) in &flashes {
            self.rows[y][x] = 0;
        }
        flashes.len()
    }

    fn incr(&mut self, x: usize, y: usize, flashes: &HashSet<Pos>) -> bool {
        if self.out_bounds(x, y) {
            return false;
        }
        self.rows[y][x] += 1;
        if self.rows[y][x] > 9 {
            !flashes.contains(&Pos(x, y))
        } else {
            false
        }
    }

    fn out_bounds(&self, x: usize, y: usize) -> bool {
        x >= self.w || y >= self.h
    }
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let mut grid = Grid::new(
        reader
            .lines()
            .flat_map(|i| i)
            .map(|i| {
                i.as_bytes()
                    .into_iter()
                    .map(|&j| j as i32 - b'0' as i32)
                    .collect()
            })
            .collect(),
    );
    println!(
        "Part 1: {}",
        (0..100).into_iter().map(|_| grid.step()).sum::<usize>()
    );
    let total = grid.w * grid.h;
    for i in 101.. {
        if grid.step() == total {
            println!("Part 2: {}", i);
            break;
        }
    }
    Ok(())
}
